text/0497-glue-l2-construct.md

```diff
@@ -0,0 +1,434 @@
+
+## Working Backwards - README
+
+[AWS Glue](https://aws.amazon.com/glue/) is a serverless data integration
+service that makes it easier to discover, prepare, move, and integrate data
+from multiple sources for analytics, machine learning (ML), and application
+development. Glue was released on 2017/08.
+[Launch](https://aws.amazon.com/blogs/aws/launch-aws-glue-now-generally-available/)
+
+Today, customers define Glue data sources, connections, jobs, and workflows
+to define their data and ETL solutions via the AWS console, the AWS CLI, and
+Infrastructure as Code tools like CloudFormation and the CDK. However, they
+have challenges defining the required and optional parameters depending on
+job type, networking constraints for data source connections, secrets for
+JDBC connections, and least-privilege IAM Roles and Policies. We will build
+convenience methods working backwards from common use cases and default to
+recommended best practices.
+
+This RFC proposes updates to the L2 construct for Glue which will provide
+convenience features and abstractions for the existing
+[L1 (CloudFormation) Constructs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/AWS_Glue.html) building on the
+functionality already supported in the [@aws-cdk/aws-glue-alpha module](https://github.com/aws/aws-cdk/blob/v2.51.1/packages/%40aws-cdk/aws-glue/README.md).
+
+## Create a Glue Job
+
+The glue-alpha-module already supports three of the four common types of Glue
+Jobs: Spark (ETL and Streaming), Python Shell, Ray. This RFC will add the
+more recent Flex Job. The construct also implements AWS practice
+recommendations when creating a Glue Job such use of Secrets Management for
+Connection JDBC strings, Glue Job Autoscaling, least privileges in terms of
+IAM permissions and also sane defaults for Glue job specification (more details
+are mentioned in the below table).
+
+This RFC will introduce breaking changes to the existing glue-alpha-module to
+streamline the developer experience and introduce new constants and validations.
+As an opinionated construct, the Glue L2 construct will enforce
+best practices and not allow developers to create resources that use deprecated
+libraries and tool sets (e.g. deprecated versions of Python).
+
+### Spark Jobs
+
+1. **ETL Jobs**
+
+ETL jobs supports python and Scala language. ETL job type supports G1, G2, G4
+and G8 worker type default as G2, which customer can override. It wil default to
+the best practice version of ETL 4.0, but allow developers to override to 3.0.
+We will also default to best practice enablement the following ETL features:
+`—enable-metrics, —enable-spark-ui, —enable-continuous-cloudwatch-log.`
+You can find more details about version, worker type and other features in
+[Glue's public documentation](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-jobs-job.html).
+
+```
+glue.ScalaSparkEtlJob(this, 'ScalaSparkEtlJob', {
```

Comment:
I don't understand the benefit of having one construct per combination of (language, type, subtype). From what I've seen in the `aws-glue-alpha` readme, `JobExecutable` already takes care of these combinations.

Reply:
We initially were going to use a factory pattern with separate sub-classes to decide based on the parameters provided, but that puts the onus on the developer to remember what parameters are required and optional for each job type. Kendra preferred having separate methods to instantiate each permutation of valid job type and its associated parameters, which enforces the contract proactively and simplifies the developer experience. The actual objects created will all be JobExecutables, not sub-classes or sub-constructs.